### PR TITLE
Routers & Security-Groups

### PR DESCRIPTION
**Pull request by afonsomatos**
Added Routers, Security-Groups and Security-Group-Roles services and models.
I left integration tests to the end because that's what made sense, since we first want to agree on a pleasant programming interface  to use.

Here are the implemented and missing services:
```
Neutron (Network):

[Simão]

     GET   /v2.0/quotas                        List quotas for projects with non-default quota values
     GET   /v2.0/quotas/{project_id}           List quotas for a project
     PUT   /v2.0/quotas/{project_id}           Update quota for a project
  DELETE   /v2.0/quotas/{project_id}           Reset quota for a project
     GET   /v2.0/quotas/{project_id}/default   List default quotas for a project

[Me]

X    GET   /v2.0/security-groups                       List security groups
X   POST   /v2.0/security-groups                       Create security group
X    GET   /v2.0/security-groups/{security_group_id}   Show security group
X    PUT   /v2.0/security-groups/{security_group_id}   Update security group
X DELETE   /v2.0/security-groups/{security_group_id}   Delete security group

X    GET   /v2.0/security-group-rules                            List security group rules
X   POST   /v2.0/security-group-rules                            Create security group rule
X    GET   /v2.0/security-group-rules/{security_group_rule_id}   Show security group rule
X DELETE   /v2.0/security-group-rules/{security_group_rule_id}   Delete security group rule

X    GET   /v2.0/networks/{network_id}    Show network details
X    PUT   /v2.0/networks/{network_id}    Update network
X DELETE   /v2.0/networks/{network_id}    Delete network
X    GET   /v2.0/networks                 List networks
X       - Lots of query parameters can be made into methods
X   POST   /v2.0/networks                 Create network
X   POST   /v2.0/networks                 Bulk Create network

X    GET   /v2.0/ports/{port_id}  Show port details
X    PUT   /v2.0/ports/{port_id}  Update port
X DELETE   /v2.0/ports/{port_id}  Delete port
X    GET   /v2.0/ports            List ports
X   POST   /v2.0/ports            Create port

X    GET   /v2.0/floatingips                   List floating IPs
X   POST   /v2.0/floatingips                   Create floating IP
X    GET   /v2.0/floatingips/{floatingip_id}   Show floating IP details
X    PUT   /v2.0/floatingips/{floatingip_id}   Update floating IP
X DELETE   /v2.0/floatingips/{floatingip_id}   Delete floating IP

X    GET   /v2.0/routers                List routers
X   POST   /v2.0/routers                Create router
X    GET   /v2.0/routers/{router_id}    Show router details
X    PUT   /v2.0/routers/{router_id}    Update router
X DELETE   /v2.0/routers/{router_id}    Delete router

X    PUT   /v2.0/routers/{router_id}/add_router_interface       Add interface to router
X    PUT   /v2.0/routers/{router_id}/remove_router_interface    Remove interface from router
X    PUT   /v2.0/routers/{router_id}/add_extraroutes            Add extra routes to router
X    PUT   /v2.0/routers/{router_id}/remove_extraroutes         Remove extra routes from router

X     GET   /v2.0/subnets               List subnets
X    POST   /v2.0/subnets               Create subnet
X    POST   /v2.0/subnets               Bulk create subnet
X     GET   /v2.0/subnets/{subnet_id}   Show subnet details
X     PUT   /v2.0/subnets/{subnet_id}   Update subnet
X  DELETE   /v2.0/subnets/{subnet_id}   Delete subnet
```

### CODE REVIEW

File: src/main/scala/pt/tecnico/dsi/neutron/NeutronClient.scala [hunk 1]
```diff
@@ -4,16 +4,23 @@ import cats.effect.Sync
 import io.circe.{Decoder, Encoder}
 import org.http4s.client.Client
 import org.http4s.{Header, Uri}
-import pt.tecnico.dsi.neutron.models.{Model, Network, Port, Subnet}
+import pt.tecnico.dsi.neutron.models.{FloatingIp, Model, Network, Port, SecurityGroup, Subnet}
 import pt.tecnico.dsi.neutron.services._
 
 class NeutronClient[F[_]: Sync](baseUri: Uri, authToken: Header)(implicit client: Client[F]) {
   val uri: Uri = baseUri / "v2.0"
```
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/NeutronClient.scala:9]: Aposto que estas 3 linhas não compilam.

File: src/main/scala/pt/tecnico/dsi/neutron/NeutronClient.scala [hunk 2]
```diff
@@ -4,16 +4,23 @@ import cats.effect.Sync
...
 
-  val networks: CrudService[F, Network] = crudService[Network]("network")
-  val subnets: CrudService[F, Subnet] = crudService[Subnet]("subnet")
-  val ports: CrudService[F, Port] = crudService[Port]("port")
+  val floatingIps: CrudService[F, FloatingIp] =
+    new CrudService[F, FloatingIp](uri, "floatingip", authToken) {}
+
+  val securityGroups: CrudService[F, SecurityGroup] =
+    new CrudService[F, SecurityGroup](uri, "security-group", authToken) {}
+
```
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/NeutronClient.scala]: Aqui dá para ver aquela inconsistência que tinha falado. O SecurityGroupRules e Routers têm uma classe service dedicada os restantes não.

File: src/main/scala/pt/tecnico/dsi/neutron/NeutronClient.scala [hunk 3]
```diff
@@ -4,16 +4,23 @@ import cats.effect.Sync
 import io.circe.{Decoder, Encoder}
 import org.http4s.client.Client
 import org.http4s.{Header, Uri}
-import pt.tecnico.dsi.neutron.models.{Model, Network, Port, Subnet}
+import pt.tecnico.dsi.neutron.models.{FloatingIp, Model, Network, Port, SecurityGroup, Subnet}
 import pt.tecnico.dsi.neutron.services._
 
 class NeutronClient[F[_]: Sync](baseUri: Uri, authToken: Header)(implicit client: Client[F]) {
   val uri: Uri = baseUri / "v2.0"
```
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/NeutronClient.scala]: Vais precisar de ter isto para conseguir correr testes.
```scala
val uri: Uri = if (baseUri.path.endsWith("v2.0") || baseUri.path.endsWith("v2.0/")) baseUri else baseUri / "v2.0"
```

File: src/main/scala/pt/tecnico/dsi/neutron/models/FloatingIp.scala [hunk 4]
```diff
@@ -0,0 +1,64 @@
+package pt.tecnico.dsi.neutron.models
+
+import java.time.LocalDateTime
+
+import io.circe.{Decoder, Encoder}
+import io.circe.derivation.{deriveDecoder, renaming, deriveEncoder}
+
+object FloatingIp {
+
```
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/models/FloatingIp.scala:10]: Mete o companion object por cima da companion classe correspondente.

File: src/main/scala/pt/tecnico/dsi/neutron/models/FloatingIp.scala [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  object Update {
+    implicit val decoder: Encoder[Update] = deriveEncoder(renaming.snakeCase)
+  }
+
+  object Create {
+    implicit val decoder: Encoder[Create] = deriveEncoder(renaming.snakeCase)
+  }
+
+  case class Create(
```
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/models/FloatingIp.scala]: Não devia ter tudo default value None?

File: src/main/scala/pt/tecnico/dsi/neutron/models/FloatingIp.scala [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    portId: Option[String],
+    projectId: Option[String],
+    fixedIpAddress: Option[String],
```
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/models/FloatingIp.scala]: `fixedIpAddress` e `floatingIpAddress` deviam ser do tipo `InetAddress`.

Consegues fazer um decoder usando a função `InetAddress.getByName`.

File: src/main/scala/pt/tecnico/dsi/neutron/models/FloatingIp.scala [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    projectId: String,
+    createdAt: LocalDateTime,
+    updatedAt: LocalDateTime,
+    revisionNumber: Integer,
+    floatingNetworkId: String,
+    fixedIpAddress: String,
+    floatingIpAddress: String,
+    portId: String,
+    tags: List[String],
+    portForwardings: List[Map[String, String]]
```
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/models/FloatingIp.scala]: O tipo de `portForwardings` parece-me errado.

File: src/main/scala/pt/tecnico/dsi/neutron/models/FloatingIp.scala [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    floatingNetworkId: Option[String]
+  )
+
+  case class Update(
+    portId: String,
+    fixedIpAddress: Option[String],
+    description: Option[String]
+  )
+
+  case class Read(
```
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/models/FloatingIp.scala:26]: Não gosto muito desta classe se chamar `Read`.

File: src/main/scala/pt/tecnico/dsi/neutron/models/Network.scala [hunk 9]
```diff
@@ -8,19 +8,21 @@ import io.circe.{Decoder, Encoder, HCursor}
...
-      routerExternal = routerExternal
-    )
+  object Read {
+    implicit val decoder: Decoder[Read] = (c: HCursor) => {
+      for {
+        providerNetworkType      <- c.downField("provider:network_type").as[String]
+        providerPhysicalNetwork  <- c.downField("provider:physical_network").as[String]
+        providerSegmentationId   <- c.downField("provider:segmentation_id").as[Integer]
+        routerExternal           <- c.downField("router:external").as[Boolean]
+        network                  <- c.as[Read](deriveDecoder(renaming.snakeCase))
```
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/models/Network.scala]: Não parece que isto vá funcionar.

File: src/main/scala/pt/tecnico/dsi/neutron/models/Route.scala [hunk 10]
```diff
@@ -0,0 +1,11 @@
+package pt.tecnico.dsi.neutron.models
+
+import io.circe.{Codec, Decoder, Encoder}
+import io.circe.derivation.{deriveEncoder, deriveDecoder, renaming}
+
+object Route {
+  implicit val encoder: Encoder[Route] = deriveEncoder(renaming.snakeCase)
```
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/models/Route.scala]: Não podias fazer o codec directamente?

File: src/main/scala/pt/tecnico/dsi/neutron/models/Router.scala [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    projectId: String,
+    name: String,
+    description: String,
+    adminStateUp: Boolean,
+    status: String, // ???
+    externalGatewayInfo: Map[String, String], // ???
+    revisionNumber: Integer,
+    routes: List[Map[String, String]], // ???
+    destination: String,
+    nexthop: String, // Ipv4Address
```
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/models/Router.scala]: Garantidamente IPv4? Se sim podes usar `Inet4Address`.

File: src/main/scala/pt/tecnico/dsi/neutron/models/Router.scala [hunk 12]
```diff
@@ -0,0 +1,80 @@
+package pt.tecnico.dsi.neutron.models
+
+import java.time.LocalDateTime
+
+import io.circe.{Decoder, Encoder}
+import io.circe.derivation.{deriveDecoder, deriveEncoder, renaming, deriveCodec}
+
+object Router extends Model {
+
```
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/models/Router.scala:10]: O mesmo que disse para o FloatingIP sobre os companion objects. E não gostar da classe chamar-se Read. E não teres default values = None no `Create` e no `Update`.

File: src/main/scala/pt/tecnico/dsi/neutron/models/RouterInterface.scala [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+import io.circe.derivation.{deriveDecoder, deriveEncoder, renaming}
+import io.circe.{Decoder, Encoder}
+
+object RouterInterface {
+  implicit val decoder: Decoder[RouterInterface] = deriveDecoder(renaming.snakeCase)
+
+  object Remove {
+    implicit val encoder: Encoder[Remove] = deriveEncoder(renaming.snakeCase)
+  }
+
```
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/models/RouterInterface.scala:12]: Nones?

File: src/main/scala/pt/tecnico/dsi/neutron/models/SecurityGroup.scala [hunk 14]
```diff
@@ -0,0 +1,52 @@
+package pt.tecnico.dsi.neutron.models
+
+import java.time.LocalDateTime
+
+import io.circe.{Decoder, Encoder}
+import io.circe.derivation.{deriveDecoder, renaming, deriveEncoder}
+
+object SecurityGroup {
+
```
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/models/SecurityGroup.scala:9]: O mesmo de cima

File: src/main/scala/pt/tecnico/dsi/neutron/models/SecurityGroupRule.scala [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+
+  case class Create(
+    projectId: Option[String],
+    remoteIpPrefix: Option[String],
+    portRangeMin: Option[Integer],
+    remoteGroupId: Option[String],
+    description: Option[String],
+    direction: String,
+    protocol: Option[String],
+    ehtertype: Option[String],
```
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/models/SecurityGroupRule.scala]: Typo? não devia ser `etherType`?

File: src/main/scala/pt/tecnico/dsi/neutron/services/Routers.scala [hunk 16]
```diff
@@ -0,0 +1,15 @@
+package pt.tecnico.dsi.neutron.services
+
+import cats.effect.Sync
+import io.circe.{Decoder, Encoder}
+import org.http4s.client.Client
+import org.http4s.{Header, Uri}
+import pt.tecnico.dsi.neutron.models.Router
+
+class Routers[F[_]: Sync: Client](baseUri: Uri, authToken: Header)
+  (implicit e: Encoder[Router#Create], f: Encoder[Router#Update], g: Decoder[Router#Read])
```
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/services/Routers.scala]: Precisas destes implicitos?

**Review by Lasering — CHANGES_REQUESTED** [review]

File: src/main/scala/pt/tecnico/dsi/neutron/NeutronClient.scala (hunk 1 above)
Comment (afonsomatos) [reply src/main/scala/pt/tecnico/dsi/neutron/NeutronClient.scala:9]: Compilam, porquê?

File: src/main/scala/pt/tecnico/dsi/neutron/models/Network.scala (hunk 9 above)
Comment (afonsomatos) [reply src/main/scala/pt/tecnico/dsi/neutron/models/Network.scala]: Sugestão?

File: src/main/scala/pt/tecnico/dsi/neutron/models/FloatingIp.scala (hunk 4 above)
Comment (afonsomatos) [reply src/main/scala/pt/tecnico/dsi/neutron/models/FloatingIp.scala:10]: Não dá. Estou a usar um `sealed trait` com o nome.

File: src/main/scala/pt/tecnico/dsi/neutron/models/FloatingIp.scala (hunk 8 above)
Comment (afonsomatos) [reply src/main/scala/pt/tecnico/dsi/neutron/models/FloatingIp.scala:26]: Sugestão?

File: src/main/scala/pt/tecnico/dsi/neutron/models/Router.scala (hunk 12 above)
Comment (afonsomatos) [reply src/main/scala/pt/tecnico/dsi/neutron/models/Router.scala:10]: Nones fixed. Sugestão para o resto?

File: src/main/scala/pt/tecnico/dsi/neutron/services/Networks.scala [hunk 17]
```diff
@@ -0,0 +1,11 @@
+package pt.tecnico.dsi.neutron.services
+
+import cats.effect.Sync
+import org.http4s.{Header, Uri}
+import org.http4s.client.Client
+import pt.tecnico.dsi.neutron.models.Network
+
+final class Networks[F[_]: Sync: Client](baseUri: Uri, authToken: Header)
```
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/services/Networks.scala:8]: Codacy found an issue: [Redundant braces after class definition](https://app.codacy.com/gh/ist-dsi/scala-neutron-client/pullRequest?prid=5802728)

File: src/main/scala/pt/tecnico/dsi/neutron/services/FloatingIps.scala [hunk 18]
```diff
@@ -0,0 +1,11 @@
+package pt.tecnico.dsi.neutron.services
+
+import cats.effect.Sync
+import org.http4s.client.Client
+import org.http4s.{Header, Uri}
+import pt.tecnico.dsi.neutron.models.{FloatingIp, Subnet}
+
+final class FloatingIps[F[_]: Sync: Client](baseUri: Uri, authToken: Header)
```
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/services/FloatingIps.scala:8]: Codacy found an issue: [Redundant braces after class definition](https://app.codacy.com/gh/ist-dsi/scala-neutron-client/pullRequest?prid=5802728)

File: src/main/scala/pt/tecnico/dsi/neutron/services/SecurityGroups.scala [hunk 19]
```diff
@@ -0,0 +1,11 @@
+package pt.tecnico.dsi.neutron.services
+
+import cats.effect.Sync
+import org.http4s.client.Client
+import org.http4s.{Header, Uri}
+import pt.tecnico.dsi.neutron.models.{FloatingIp, SecurityGroup}
+
+final class SecurityGroups[F[_]: Sync: Client](baseUri: Uri, authToken: Header)
```
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/services/SecurityGroups.scala:8]: Codacy found an issue: [Redundant braces after class definition](https://app.codacy.com/gh/ist-dsi/scala-neutron-client/pullRequest?prid=5802728)

File: src/main/scala/pt/tecnico/dsi/neutron/services/Subnets.scala [hunk 20]
```diff
@@ -0,0 +1,11 @@
+package pt.tecnico.dsi.neutron.services
+
+import cats.effect.Sync
+import org.http4s.client.Client
+import org.http4s.{Header, Uri}
+import pt.tecnico.dsi.neutron.models.{Network, Subnet}
+
+final class Subnets[F[_]: Sync: Client](baseUri: Uri, authToken: Header)
```
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/services/Subnets.scala:8]: Codacy found an issue: [Redundant braces after class definition](https://app.codacy.com/gh/ist-dsi/scala-neutron-client/pullRequest?prid=5802728)

File: src/main/scala/pt/tecnico/dsi/neutron/services/Ports.scala [hunk 21]
```diff
@@ -0,0 +1,11 @@
+package pt.tecnico.dsi.neutron.services
+
+import cats.effect.Sync
+import org.http4s.client.Client
+import org.http4s.{Header, Uri}
+import pt.tecnico.dsi.neutron.models.{Port, Subnet}
+
+final class Ports[F[_]: Sync: Client](baseUri: Uri, authToken: Header)
```
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/services/Ports.scala:8]: Codacy found an issue: [Redundant braces after class definition](https://app.codacy.com/gh/ist-dsi/scala-neutron-client/pullRequest?prid=5802728)

File: src/main/scala/pt/tecnico/dsi/neutron/services/SecurityGroups.scala (hunk 19 above)
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/services/SecurityGroups.scala:8]: Codacy found an issue: [Redundant braces after class definition](https://app.codacy.com/gh/ist-dsi/scala-neutron-client/pullRequest?prid=5802728)

File: src/main/scala/pt/tecnico/dsi/neutron/services/Networks.scala (hunk 17 above)
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/services/Networks.scala:8]: Codacy found an issue: [Redundant braces after class definition](https://app.codacy.com/gh/ist-dsi/scala-neutron-client/pullRequest?prid=5802728)

File: src/main/scala/pt/tecnico/dsi/neutron/models/package.scala [hunk 22]
```diff
@@ -1,10 +1,16 @@
 package pt.tecnico.dsi.neutron
 
+import java.net.Inet4Address
```
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/models/package.scala:3]: Codacy found an issue: [Use brackets to group imports from the same package](https://app.codacy.com/gh/ist-dsi/scala-neutron-client/pullRequest?prid=5802728)

File: src/main/scala/pt/tecnico/dsi/neutron/services/Subnets.scala (hunk 20 above)
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/services/Subnets.scala:8]: Codacy found an issue: [Redundant braces after class definition](https://app.codacy.com/gh/ist-dsi/scala-neutron-client/pullRequest?prid=5802728)

File: src/main/scala/pt/tecnico/dsi/neutron/services/Ports.scala (hunk 21 above)
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/services/Ports.scala:8]: Codacy found an issue: [Redundant braces after class definition](https://app.codacy.com/gh/ist-dsi/scala-neutron-client/pullRequest?prid=5802728)

File: src/main/scala/pt/tecnico/dsi/neutron/services/FloatingIps.scala (hunk 18 above)
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/services/FloatingIps.scala:8]: Codacy found an issue: [Redundant braces after class definition](https://app.codacy.com/gh/ist-dsi/scala-neutron-client/pullRequest?prid=5802728)

Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/services/FloatingIps.scala:8]: Codacy found an issue: [Redundant braces after class definition](https://app.codacy.com/gh/ist-dsi/scala-neutron-client/pullRequest?prid=5802728)

File: src/main/scala/pt/tecnico/dsi/neutron/models/package.scala (hunk 22 above)
Comment (Lasering) [inline src/main/scala/pt/tecnico/dsi/neutron/models/package.scala:3]: Codacy found an issue: [Use brackets to group imports from the same package](https://app.codacy.com/gh/ist-dsi/scala-neutron-client/pullRequest?prid=5802728)

File: src/main/scala/pt/tecnico/dsi/neutron/models/FloatingIp.scala (hunk 4 above)
Comment (Lasering) [reply src/main/scala/pt/tecnico/dsi/neutron/models/FloatingIp.scala:10]: Tem que dar, ou não me expliquei bem, ou não estou a perceber o problema.
O que estava a dizer era:
```scala
object FloatingIp {
 object Read { ... }
 case class Read(...)

 object Update { ... }
 case class Update(...)

 object Create { ... }
 case class Create(...)
}
```

File: src/main/scala/pt/tecnico/dsi/neutron/models/FloatingIp.scala (hunk 8 above)
Comment (Lasering) [reply src/main/scala/pt/tecnico/dsi/neutron/models/FloatingIp.scala:26]: Não sei, eu percebi porque é que lhe chamaste Read, mas não gosto. Eu continuo a dizer que a class Model não está a acrescentar valor.

File: src/main/scala/pt/tecnico/dsi/neutron/models/FloatingIp.scala (hunk 4 above)
Comment (afonsomatos) [reply src/main/scala/pt/tecnico/dsi/neutron/models/FloatingIp.scala:10]: Ah ok, percebi mal.

**Review by Lasering — APPROVED** [review]

**Pull request merged**